pints/residuals_diagnostics.py

```diff
@@ -0,0 +1,255 @@
+                                   posterior_interval=0.95):
+    """
+    Generate an autocorrelation plot of the residuals.
+
+    This function can be used to analyze the results of either optimisation or
+    MCMC Bayesian inference. When multiple samples of the residuals are present
+    (corresponding to multiple MCMC samples), the plot can illustrate the
+    distribution of autocorrelations across the MCMC samples. At each lag,
+    a dot is drawn at the median autocorrelation, and a line is drawn giving
+    the extent of the posterior interval specified as an argument (by default,
+    the 2.5th to the 97.5th percentile).
+
+    When multiple outputs are present, one residuals plot will be generated for
+    each output.
+
+    Returns a ``matplotlib`` figure.
+
+    Parameters
+    ----------
+    parameters
+        The parameter values with shape ``(n_samples, n_parameters)``. When
+        passing a single best fit parameter vector, ``n_samples`` will be 1.
+    problem
+        The problem given by a :class:``pints.SingleOutputProblem`` or
+        :class:``pints.MultiOutputProblem``, with ``n_parameters`` greater than
+        or equal to the ``n_parameters`` of the `parameters`. Extra parameters
+        not found in the problem are ignored.
+    max_lag
+        Optional int value (default 10). The highest lag to plot.
+    thinning
+        Optional int value (greater than zero). If thinning is set to ``n``,
+        only every nth sample in parameters will be used. If set to ``None``
+        (default), some thinning will be applied so that about 200 samples will
+        be used.
+    significance_level
+        ``None`` or float value (default 0.05). When a significance level is
+        provided, dashed lines for the confidence interval corresponding to
+        that significance level are drawn on the plot. When ``None``, no lines
+        are drawn.
+    posterior_interval
+        Float value (default 0.95). When multiple samples of the parameter
+        values are provided, this gives the size of the credible region of the
+        posterior to plot.
+    """
+    import matplotlib.pyplot as plt
+    import numpy as np
+    import scipy.special
+
+    # Get the number of problem outputs
+    n_outputs = problem.n_outputs()
+
+    # Get the matrix of residuals values
+    residuals = calculate_residuals(parameters, problem, thinning=thinning)
+
+    # Get the number of samples
+    n_samples = residuals.shape[0]
+
+    # If there are multiple samples of the parameters, calculate the
+    # percentiles of the posterior to plot
+    if n_samples > 1:
+        if posterior_interval > 1 or posterior_interval < 0:
+            raise ValueError('posterior interval must fall between 0 and 1')
+        upper_pctle = (0.5 + posterior_interval / 2) * 100
+        lower_pctle = (0.5 - posterior_interval / 2) * 100
+
+    # Find the location of the confidence interval dotted lines
```

Comment:
Do we have a citation for the confidence intervals? If so, can we include it in the docstrings? Alternatively / additionally, it may be worth giving the mathematical form of this in the docstrings for this method.

Reply:
I added a citation which I think is applicable here. I also added the basic equations to the docstring.

Reply:
Looks good!